Feat/render model

ajatdarojat45 opened the pull request:
This pull request introduces major new features for 3D model simulation and visualization, including a Three.js-powered viewport and support for extracting and rendering Rhino 3DM files. It also adds documentation and links for multiple simulation methods, refines the application layout, and expands the project's dependencies to support these new capabilities.

**3D Model Visualization and Rendering:**

* Added new components (`ModelViewer`, `ViewportCanvas`, `ModelRenderer`) to enable interactive 3D model viewing using react-three-fiber and drei, including camera controls, grid, axes, and error/loading states. Models are loaded and rendered from URLs and support switching between perspective and orthographic cameras. [[1]](diffhunk://#diff-42dbf7f948aa568cd6412cf3a6f4fc75b5f5f2a42ea5df40901f1f259d531bd3R1-R43) [[2]](diffhunk://#diff-29c5bbfad94061c245592b3067f5fe0d43ff8f2599c10da319e044fc48f2fe1bR1-R113) [[3]](diffhunk://#diff-b5b620cbd161ceb70c185b902c705f0243d7e2e81c2c20ff769adbd6a89c8e6dR1-R17)
* Implemented helper modules for extracting Rhino 3DM files from ZIP archives, parsing them, processing layers, and optimizing models for rendering. This includes functions for file extraction (`fileExtractor.ts`), layer processing (`layerProcessor.ts`), and model processing (`modelProcessor.ts`). [[1]](diffhunk://#diff-3fdb52efc04fdcadf163897c5c6cb50a7da3b8d668a488c8adc33ea6ecdc278dR1-R81) [[2]](diffhunk://#diff-2c655f1c64ee6c72e3165788d2e747422b477b3ab25e11a18ab2772e1928ab28R1-R147) [[3]](diffhunk://#diff-971defdd37e8095e9bbfefaf714d33af52ead113d053093d42220bacc76a5d7bR1-R129)

**Simulation Method Documentation:**

* Added a new constant `SIMULATION_METHODS` containing names, documentation, and GitHub links for various simulation approaches, making it easier to reference and display simulation options.

**Dependency Updates:**

* Updated `package.json` to version 0.0.6 and added dependencies for Three.js, react-three-fiber, drei, jszip, and type definitions needed for 3D rendering and file extraction. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L4-R4) [[2]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R34-R43) [[3]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L49-R56)

**UI/Layout Improvements:**

* Improved the main application layout to ensure the header does not shrink and the main content area (including the viewport) uses available space with proper overflow handling.

**Changelog Update:**

* Documented all new features, bug fixes, and improvements in `CHANGELOG.md` for version 0.0.6.

Commit:
feat: create 3d viewport with helper

Commit:
feat: extract file from modelUrl to be parse with rhino3dm and render

Commit:
fix: change open model button to redirect to editor page

Commit:
chore: format changes based on new prettier config

Commit:
fix: conflict before merging

Commit:
fix: create copy of file that being parsed because of conflict

Commit:
fix: grid visibility from below angle

Commit:
fix: model caching not active during rendering

ajatdarojat45 committed:
Merge pull request #10 from tigadika/feat/3dviewport

feat: 3DViewport

ajatdarojat45 committed:
feat: add simulation methods with relevant links and documentation

ajatdarojat45 committed:
chore(release): 0.0.6

ajatdarojat45 commented:
Hi @SilvinWillemsen,
This PR includes render 3d model.
Looking forward to your feedback and review. Thanks!

SilvinWillemsen commented on src/constants/index.ts:
```diff
@@ -5,3 +5,41 @@ export const COLORS = {
+
+export const SIMULATION_METHODS = {
+  DE: {
+    name: "Differential Equations",
```
*Diffusion Equation

SilvinWillemsen commented:
Hi @ajatdarojat45, the viewport is looking good! A few remarks (also see [video](https://github.com/user-attachments/assets/5d8db10c-530d-45a5-93a0-765908440aa7):
- DE = Diffusion Equation (not Differential Equations)
- There is a white bar when you scroll down
- Rescaling the UI up resizes the viewport perfectly. The other way around, this doesn't happen. I remember this also being an issue with the previous version. Does this have something to do with Rhino?
- Why do you need the 3dm extraction from .zip? Is that because the backend provides a .zip file? 

Let me know if you have any questions!

ajatdarojat45 committed:
Change name from 'Differential Equations' to 'Diffusion Equation'

ajatdarojat45 commented:
Hi @SilvinWillemsen,

Thanks a lot for the feedback and notes! 
- Diffusion Equation — got it, We will fix the label accordingly.
- White bar on scroll — noted, We will correct that as well.
- Viewport rescaling issue — this seems related to how React Three Fiber handles the camera component. From what We’ve read in the community discussions, it’s a known issue where resizing down doesn’t properly trigger camera recalculation. It’s not directly caused by Rhino itself — Rhino3DLoader is just used for loading .3dm files (commonly from CAD models) that contain richer metadata like materials and mesh details. We will revisit this part again and see if we can find a more stable fix later on.
- 3dm extraction from .zip — yes, the backend currently provides the model file as a .zip, and inside it there’s the .3dm file. That’s why extraction is needed before loading it in the viewer.

Let me know if there’s anything else you’d like me to check or clarify!

Commit:
fix: viewport resizing by making viewport fixed, so it can ressize with window size

Commit:
fix: excess whitespace in the bottom of editor page

ajatdarojat45 committed:
Merge pull request #13 from tigadika/feat/3dviewport

Feat/3dviewport

ajatdarojat45 committed:
chore(release): 0.0.7

ajatdarojat45 committed:
Merge pull request #14 from ajatdarojat45/development

Development

ajatdarojat45 commented:
Hi @SilvinWillemsen,

We’ve fixed points 1 to 3 — please have a look when you get a chance:
- Updated label to Diffusion Equation.
- Removed the white bar on scroll.
- Adjusted the viewport scaling issue — it should now resize properly both ways.

Please let me know if everything looks good on your side!

SilvinWillemsen commented:
Looking good @ajatdarojat45! Will merge the pull request now :)
